Allow some code to be silently pre-run

jobovy opened the pull request:
Hi,

Just found this repository and it looks great! Inspired by your PRs to `numpy` and `sympy` to use this REPL to allow people to try out code in documentation pages, I was interested in using this for some of my code. However, for packages that aren't part of `pyodide` directly, users would have to first do something like
```
import micropip
await micropip.install('PACKAGE')
```
before they can `import PACKAGE`. That would be confusing for users who don't know `pyodide`. My understanding is that `pyodide` does not want to include Python packages that have pure Python wheels available on PyPI, so this would be an issue for many packages.

This quick PR allows for an `install` URL parameter that lists a set of packages to quietly install with `micropip` (I wasn't sure whether I should use `micropip` or `piplite`, but both work). It works both with lists of packages (`&install=kimmy,scipy`) or with wheel URLS (`&install=https://astro.utoronto.ca/~bovy/galpywheels/galpy-1.8.0.dev0-py3-none-any.whl&code=import%20galpy`) .

I don't have any prior experience with TypeScript or the Jupyter backend, so there might be better ways to do this, but injecting the code and removing the cell seems to do what I want (I couldn't figure out how to just have the kernel execute a bit of code...).

This works locally, but I could push a version somewhere to illustrate use better if necessary (and can add to CHANGELOG and what not if you're inclined to accept this PR).

Thanks!

Edit: Full example using my galaxy chemical evolution code [`kimmy`](https://github.com/jobovy/kimmy) on the Vercel deployment [here](https://replite-git-fork-jobovy-quiet-installs-jtpx.vercel.app/retro/consoles/?kernel=python&prerun=import%20micropip%0Aawait%20micropip.install([%27scipy%27,%27kimmy%27])&code=import%20kimmy%0Aimport%20numpy%0Aimport%20astropy.units%20as%20u%0Afrom%20matplotlib%20import%20pyplot%20as%20plt%0A%matplotlib%20inline%0Aoz=%20kimmy.OneZone()%0Ats=%20numpy.linspace(0.001,10.,1001)*u.Gyr%0Aoz.initial()%0Aplt.plot(oz.Fe_H(ts),oz.O_Fe(ts))%0Aplt.xlabel(%27%5BFe/H%5D%27)%0Aplt.ylabel(%27%5Ba/Fe%5D%27)).

jobovy committed:
Allow quiet installs of non-pyodide packages with microip

jobovy committed:
Remove window.console call

jtpio commented:
Thanks @jobovy for starting this!

jtpio commented on src/index.ts:
```diff
@@ -164,6 +165,13 @@ const parameters: JupyterFrontEndPlugin<void> = {
+      if (install) {
+        await console.sessionContext.ready;
+        console.inject('import micropip');
+        install[0].split(',').forEach(pckg => console.inject(`await micropip.install('${pckg}')`));
```
Since JupyterLite is kernel-agnostic, `replite` could also be used for other languages such as Lua (https://github.com/jupyterlite/xeus-lua-kernel) or SQLite (https://github.com/jupyterlite/xeus-sqlite-kernel).

Which means the `micropip.install()` will not be available.

Maybe an alternative would be to provide a `silent` or `prerun` option and let the users pass the `micropip.install` code in the query string parameter?

jobovy replied:
That had been my original thought (to just let users pass full code to silently run), but I thought it might be bad form for arbitrary code to be run non-transparently. I see that there is some discussion at jupyterlite/jupyterlite/issues/461 along similar lines, although I think the concerns raised there are less of an issue here. I'm happy to change the code to do this instead; it would be useful to have other run silently run (like configs) when using this to allow people to try out code in the browser.

Last night I ran into another issue, which is that I hadn't noticed that the execution counter doesn't get reset, so the first cell that is run is something like `[4]` or some other number != 1 depending on how much code was run silently. That might also be a bit confusing for users. However, even after spending a long time trawling through the jupyterlab code, I couldn't figure out how to reset the execution counter, maybe it isn't possible. Anyway, it's not a big deal that the counter doesn't start at `[1]`.

jobovy committed:
Edit if statement, because always true now

jobovy committed:
Switch to a general set of commands to pre-run silently

jobovy committed:
Run prerun code using direct kernel call so execution counter doesn't get incremented

jobovy commented:
I switched to allowing arbitrary code to pre-run silently as you suggested (with a `prerun` keyword). I also switched to invoking the kernel directly in a way that I think should not increase the execution counter (so the first displayed cell has `[1]`), but I think there's a small bug in the pyolite kernel that causes this not to work right now (see https://github.com/jupyterlite/jupyterlite/issues/478). But if that bug were fixed in pyolite, this would work perfectly.

jobovy committed:
Add example of prerun using xarray to the README

jobovy commented:
I added an example to the README (no video though) and I think this is now good to go (aside from updating jupyterlite once they release a new version, but the prompt issue is a minor one). I'm eager to start using this in my non-pyodide packages' documentation. However, I did see that you opened https://github.com/jupyterlite/jupyterlite/pull/498, so perhaps there isn't much point in working further on this here and I should just wait to open a PR there once https://github.com/jupyterlite/jupyterlite/pull/498 is merged? Or do you want to include this change in https://github.com/jupyterlite/jupyterlite/pull/498?

jtpio commented:
Thanks @jobovy.

Yes I am planning to make a new release once https://github.com/jupyterlite/jupyterlite/pull/498 lands. Which means this repo will not be strictly necessary since every JupyterLite deployment will have it by default.

We should indeed consider allowing running code on startup: https://github.com/jupyterlite/jupyterlite/issues/461. It's not clear yet whether this should be available by default (w.r.t reproducibility), but it can be developed as an extension (or here in the `replite` repo)

jobovy committed:
Upgrade to jupyterlite version that correctly honors execute_request's store_history

jobovy commented:
I'm going to close this, because this PR is irrelevant now that this repository just uses the `jupyterlite` repl app (so the issue should really be solved there). Thanks for your help in getting this to work!
